vignettes/delta-method.Rmd

```diff
@@ -110,3 +119,209 @@ Take the second derivative by subtracting each of the levels from the base level
+\end{align*}
+\ldots which is the first term of our Jacobian.^[Note that the first part of the expression is sometimes written $\frac{1}{1 + \exp(-X_i\beta)} \cdot \frac{1}{1 + \exp(X_i\beta)}$ since that implies $\frac{\partial}{\partial \beta} \text{logit}^{-1}(X\beta) = \text{logit}^{-1}(X\beta) \cdot \text{logit}^{-1}(-X\beta)$]
+
+What about the other terms? Well, because the $X_i\beta$ term is additively separable, each term is the same, except for having a different term at the end - for example, the derivative with respect to $\beta_2$ is going to be $\frac{1}{n} \sum_{i = 1}^n \frac{\exp(-X_i\beta))}{(1 + \exp(-X_i\beta))^2} \cdot \text{age}_i$. So the full Jacobian is
+\[
+J = 
+\left[
+  \begin{array}{l} \\
+    \frac{1}{n} \sum_{i = 1}^n \frac{\exp(-X_i\beta)}{(1 + \exp(-X_i\beta))^2} \cdot 1\\
+    \frac{1}{n} \sum_{i = 1}^n \frac{\exp(-X_i\beta)}{(1 + \exp(-X_i\beta))^2} \cdot \tau_i\\
+    \frac{1}{n} \sum_{i = 1}^n \frac{\exp(-X_i\beta)}{(1 + \exp(-X_i\beta))^2} \cdot \text{age}_i\\
+    \frac{1}{n} \sum_{i = 1}^n \frac{\exp(-X_i\beta)}{(1 + \exp(-X_i\beta))^2} \cdot \tau_i \cdot \text{age}_i
+  \end{array}
+\right]
+\]
+
+This can be rewritten as
+
+\[
+J = 
+\frac{1}{n}
+\left[
+  \begin{array} \\
+     \frac{\exp(-X_1\beta)}{(1 + \exp(-X_1\beta))^2} &
+     \frac{\exp(-X_2\beta)}{(1 + \exp(-X_2\beta))^2} &
+     \cdots &
+     \frac{\exp(-X_n\beta)}{(1 + \exp(-X_n\beta))^2}
+  \end{array}
+\right]
+\left[
+  \begin{array}{cccc}\\
+    1 & \tau_1 & \text{age}_1 & \tau_1 \cdot \text{age}_1 \\
+    1 & \tau_2 & \text{age}_2 & \tau_2 \cdot \text{age}_2 \\
+    \vdots  & \vdots & \vdots & \vdots\\
+    1 & \tau_n & \text{age}_n & \tau_n \cdot \text{age}_n \\
+  \end{array}
+\right]
+\]
+
+The first term applies the derivative of the inverse link function to every linear predictor in the model, and the second term is the covariate matrix from our data! How convenient! That's going to be true for all general linear models. So we can rewrite this as
+\[J = \frac{\left[\begin{array}\\ \frac{\exp(-X_1\beta)}{(1 + \exp(-X_1\beta))^2} & \frac{\exp(-X_2\beta)}{(1 + \exp(-X_2\beta))^2} & \cdots & \frac{\exp(-X_n\beta)}{(1 + \exp(-X_n\beta))^2}\end{array}\right] X}{n}\]
+Recalling that our variance is $J V J^T$, and that we've already calculated $X_i\beta$ for all $i$, this is relatively simple to do in R:
+
+```{r}
+# Get the data
+x <- model.matrix(lg)
+
+# CONTROL ERROR
+
+# Apply the derivative of the inverse link function to the linear predictors
+deriv <- as.vector(exp(-pred_ctl) / (1 + exp(-pred_ctl))^2)
+
+# Set treatment to 0
+x[, 'treatment'] <- 0
+x[, 'treatment:age'] <- x[, 'treatment'] * x[, 'age']
+
+# Complete the chain rule by matrix-multiplying the derivatives by the data,
+# now we have the jacobian
+j <- deriv %*% x / nrow(x)
+
+# The variance of our estimate is the cross product of the jacobian and the model's
+# variance-covariance matrix
+variance <- j %*% vcov(lg) %*% t(j)
+
```

Comment:
I just realized that I'm not doing this in the more efficient way that the package and explanation above do - I should be stacking the `deriv` objects from treatment = 0 and treatment = 1, and then do one matrix operation / get a vector back instead of two totally separate calculations / get two scalars back. Worth updating?